apps/src/lib/tools/jsdebugger/redux.js

```diff
@@ -0,0 +1,261 @@
+
+function getRunApp(state) {
+  return getRoot(state).runApp;
+}
+
+export function isPaused(state) {
+  return getRoot(state).jsInterpreterPaused;
+}
+
+function getObserver(state) {
+  return getRoot(state).observer;
+}
+
+function getWatchIntervalId(state) {
+  return getRoot(state).watchIntervalId;
+}
+
+export function isAttached(state) {
+  return !!getJSInterpreter(state);
+}
+
+export function canRunNext(state) {
+  return (
+    isPaused(state) &&
+    getRoot(state).jsInterpreterNextStep === JSInterpreter.StepType.RUN
+  );
+}
+
+export function getLogOutput(state) {
+  return getRoot(state).logOutput;
+}
+
+export const selectors = {
+  getRoot,
+  getCommandHistory,
+  getJSInterpreter,
+  isPaused,
+  isAttached,
+  canRunNext,
+  getLogOutput,
+};
+
+// actions
+export function initialize({runApp}) {
+  return {type: INITIALIZE, runApp};
+}
+
+export function appendLog(output) {
+  return {type: APPEND_LOG, output};
+}
+
+export function clearLog() {
+  return {type: CLEAR_LOG};
+}
+
+function mirrorInterpreterState(jsInterpreter) {
+  return {type: MIRROR_INTERPRETER_STATE, jsInterpreter};
+}
+
+export function attach(jsInterpreter) {
+  return (dispatch, getState) => {
+    const observer = new Observer();
+    observer.observe(
+      jsInterpreter.onNextStepChanged,
+      () => dispatch(mirrorInterpreterState(jsInterpreter))
+    );
+    observer.observe(
+      jsInterpreter.onPause,
+      () => dispatch(togglePause())
+    );
+    observer.observe(
+      jsInterpreter.onExecutionWarning,
+      (output) => dispatch(appendLog(output))
+    );
+
+    const watchIntervalId = setInterval(
+      () => {
+        const jsInterpreter = getJSInterpreter(getState());
+        getState().watchedExpressions.forEach(we => {
+          const currentValue = jsInterpreter.evaluateWatchExpression(we.get('expression'));
+          dispatch(updateWatchExpressions(we.get('expression'), currentValue));
+        });
+      },
+      WATCH_TIMER_PERIOD
+    );
+    dispatch(clearLog());
+    dispatch({
+      type: ATTACH,
+      observer,
+      watchIntervalId,
+      jsInterpreter,
+    });
+  };
+}
+
+export function detach() {
+  return (dispatch, getState) => {
+    const observer = getObserver(getState());
+    observer && observer.unobserveAll();
+    clearInterval(getWatchIntervalId(getState()));
+    dispatch({type: DETACH});
+  };
+}
+
+export function togglePause() {
+  return (dispatch, getState) => {
+    const jsInterpreter = getJSInterpreter(getState());
+    if (!jsInterpreter) {
+      return;
+    }
+    jsInterpreter.handlePauseContinue();
+    dispatch(mirrorInterpreterState(jsInterpreter));
+  };
+}
+
+export function stepIn() {
+  return (dispatch, getState) => {
+    let jsInterpreter = getJSInterpreter(getState());
+    if (!jsInterpreter) {
+      const runApp = getRunApp(getState());
+      if (runApp) {
+        runApp();
+      } else {
+        throw new Error("jsdebugger has not been initialized yet");
+      }
```

Comment:
I don't like that this code depends on `runApp` dispatching the `attach()` action with an interpreter. Seems kind of upside down. I wonder if the `<JsDebugger/>` react component should be responsible for starting program execution before trying to step in.